p2pserver/link/link.go

```diff
@@ -166,44 +198,75 @@ func (this *Link) CloseConn() {
 	}
 }
 
-func (this *Link) Send(msg types.Message) error {
-	sink := comm.NewZeroCopySink(nil)
-	types.WriteMessage(sink, msg)
-
-	return this.SendRaw(sink.Bytes())
-}
-
-func (this *Link) SendRaw(rawPacket []byte) error {
+func (this *Link) writeLoop() {
 	conn := this.conn
 	if conn == nil {
-		return errors.New("[p2p]tx link invalid")
+		return
 	}
 
-	nByteCnt := len(rawPacket)
-	log.Tracef("[p2p]TX buf length: %d\n", nByteCnt)
+	for {
+		this.lock.Lock()
+		for this.err == nil && this.writeBuff.Size() == 0 {
+			this.cond.Wait()
+		}
+		if this.err != nil {
+			this.lock.Unlock()
+			return
+		}
+		this.writeBuff.Reset()
+		temp := this.writeBuff
+		this.writeBuff = this.pendingBuff
+		this.pendingBuff = temp
+		this.lock.Unlock()
+
+		payload := this.writeBuff.Bytes()
+		nByteCnt := len(payload)
+		log.Tracef("[p2p]TX buf length: %d\n", nByteCnt)
+
+		nCount := nByteCnt / common.PER_SEND_LEN
+		if nCount == 0 {
+			nCount = 1
+		}
 
-	nCount := nByteCnt / common.PER_SEND_LEN
-	if nCount == 0 {
-		nCount = 1
-	}
-	conn.SetWriteDeadline(time.Now().Add(time.Duration(nCount*common.WRITE_DEADLINE) * time.Second))
-	_, err := conn.Write(rawPacket)
-	if err != nil {
-		log.Infof("[p2p]error sending messge to %s :%s", this.GetAddr(), err.Error())
-		this.disconnectNotify()
-		return err
+		conn.SetWriteDeadline(time.Now().Add(time.Duration(nCount*common.WRITE_DEADLINE) * time.Second))
+		_, err := conn.Write(payload)
+		if err != nil {
+			this.lock.Lock()
+			this.err = err
+			this.disconnectNotify()
+			this.cond.Signal() // notify read loop
```

Comment:
signal after unlock?

and make broadcast in disconnectNotify?

Reply:
there are just `readLoop` and `writeLoop` using this condition variable, since we are in `writeLoop`, only `readLoop` will wait for this, so `Signal` should be enough.

Reply:
from go doc, it is ok to call `Singal` in locked context.